### PR TITLE
caf: Base skeleton for Common Application Framework.

### PR DESCRIPTION
**Pull request by emob-nordic**
Scheleton for CAF with base module (`module_state_event`) and `buttons` moved to `subsys/caf`.

### CODE REVIEW

**Comment by pdunaj** [comment]
Hi @carlescufi , is the location fine with you? It could be in applications/caf but subsystems also looks nice.

**Comment by carlescufi** [comment]
Adding @nordic-krch and @pabigot since they have worked on low-level frameworks upstream that could bear some resemblance to CAF.

**Comment by carlescufi** [comment]
> Hi @carlescufi , is the location fine with you? It could be in applications/caf but subsystems also looks nice.

`subsys/caf` is a good location, thanks for checking @pdunaj.

File: subsys/caf/modules/buttons.c
Comment (pabigot) [inline subsys/caf/modules/buttons.c:105]: It looks like this code isn't respecting the devicetree flags that specify whether the logical signal is active high or active low.  This may confuse developers who configure that level in devicetree.

**Review by pabigot — COMMENTED** [review]
It seems this is mostly refactoring and renaming of existing code; unless reviewing that code is in scope I have no position on whether this is good or bad.

I am concerned that GPIO signal polarity is being defined globally via Kconfig rather than specified on a per-gpio basis with devicetree flags like `GPIO_ACTIVE_LOW`, but that too is pre-existing behavior.

File: subsys/caf/modules/buttons.c
Comment (pdunaj) [reply subsys/caf/modules/buttons.c:105]: Hi @pabigot , thanks for having a look. You are right that this is essentially moving the code from location A to B. Refactoring is done to make it usable by other applications that are on the way.

The initial goal is to move it to reuse the code. Due to timelines we have I would like to avoid big changes (if making it reusable will take too long we will not be able to reuse it for the apps we work on).
However after code is made common we would like it to fit in various cases. And for that reason we will need to plan further refactoring, updating and improving. For nrf_desktop alone there would be no point to do it.

> It looks like this code isn't respecting the devicetree flags that specify whether the logical signal is active high or active low. This may confuse developers who configure that level in devicetree.
That is true. When it was written there was no way of pass the gpio flags in dt. This is one of the limitations that we will need to fix. I would prefer to fix it in the following changes. We can do it in parallel to app development.

In the nrf_desktop there are several modules that are potentially transferable to caf (I have selected these two as I find them simplest and also because we will find the use for them rightaway). When moving the modules we will need to identify nrf_desktop specifics and get rid of them. Secondly we will need to find things that are outdated due to zephyr moving foreward and update them. We won't be able however to fix everything at once.

**Comment by pdunaj** [comment]
This draft is actually a mess. @emob-nordic is preparing separate PRs for each module.

**Review by greg-fer — CHANGES_REQUESTED** [review]
Concerning documentation, I won't be able to review before NCS 1.5.0 release, but what I can say now is that the `subsys` directory should not contain doc `.rst` files. The documentation for subsys libraries should go to `nrf/doc/nrf/libraries/` into a new `caf` directory (or similar), and in `nrf/libraries` we need an introductory page that describes the purpose, use cases, and common basic configuration for these modules.

File: subsys/caf/events/CMakeLists.txt
Comment (tejlmand) [inline subsys/caf/events/CMakeLists.txt:7]: Why are we added files to the `app` library ?

The `app` library is the users library, and solely the user should decide what goes in there.

File: applications/nrf_desktop/configuration/nrf52820dongle_nrf52820/app_ZReleaseMCUBoot.conf
Comment (tejlmand) [inline applications/nrf_desktop/configuration/nrf52820dongle_nrf52820/app_ZReleaseMCUBoot.conf:9]: Just a small nit.
Maybe move this line a little down, so that all `CAF` settings are grouped together like:
```
CONFIG_CAF=y
CONFIG_CAF_BUTTONS_ENABLE=y
CONFIG_CAF_BUTTONS_PM_EVENTS_ENABLE=y
CONFIG_CAF_BUTTONS_POLARITY_INVERSED=y
```

This applies to all conf files.

**Review by tejlmand — CHANGES_REQUESTED** [review]

File: applications/nrf_desktop/configuration/nrf52820dongle_nrf52820/app_ZReleaseMCUBoot.conf
Comment (emob-nordic) [reply applications/nrf_desktop/configuration/nrf52820dongle_nrf52820/app_ZReleaseMCUBoot.conf:9]: Will do. Just need a bit of a clean up rn.

File: subsys/caf/events/CMakeLists.txt
Comment (emob-nordic) [reply subsys/caf/events/CMakeLists.txt]: That's right. Changed to `zephyr_sources(_ifdef)`,similar to other subsystems like `shell`.

**Review by greg-fer — CHANGES_REQUESTED** [review]
As mentioned during the f2f call with Emil and Pawel Dunaj, we need to keep the nRF Desktop button module rst file rather than remove it, and add there the nRF Desktop event table.

**Review by tejlmand — APPROVED** [review]

**Comment by emob-nordic** [comment]
Without last commit I'm getting:
```
-:1929: WARNING:FUNCTION_ARGUMENTS: function definition argument '__module_' should also have an identifier name
#1929: FILE: subsys/caf/events/module_state_event.h:57:
+extern const void * const _CONCAT(__module_, MODULE);

-:1929: WARNING:FUNCTION_ARGUMENTS: function definition argument 'MODULE' should also have an identifier name
#1929: FILE: subsys/caf/events/module_state_event.h:57:
+extern const void * const _CONCAT(__module_, MODULE);
```
I have no better idea than this in the PR to resolve that issue. @carlescufi  any ideas?

**Comment by wiminordic** [comment]
@pdunaj this PR causes Keyboard to crash on startup:
[00000311] <inf> event_manager: e: button_event key_id=0x300 pressedASSERTION FAIL [0] @ WEST_TOPDIR/nrf/applications/nrf_desktop/src/modules/hid_state.c:1450
[00000317] <err> os: r0/a1:  0x00000004  r1/a2:  0x000005aa  r2/a3:  0x200027c0
[00000317] <err> os: r3/a4:  0x00000014 r12/ip:  0x34313a63 r14/lr:  0x000174ed
[00000317] <err> os:  xpsr:  0x61000000
[00000317] <err> os: Faulting instruction address (r15/pc): 0x0002faa6
[00000318] <err> os: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0
[00000318] <err> os: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0
[00000318] <err> os: Current thread: 0x200030e0 (unknown)
[00611434] <err> os: Halting system

**Comment by emob-nordic** [comment]
> [00000317] os: Faulting instruction address (r15/pc): 0x0002faa6
> [00000318] os: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0
> [00000318] os: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0
> [00000318] os: Current thread: 0x200030e0 (unknown)
> [00611434] os: Halting system

Will fix it in a minute.

**Comment by emob-nordic** [comment]
@wiminordic It should be fine, thanks for reporting. I missed that one..

File: applications/nrf_desktop/src/hw_interface/Kconfig.buttons_sim
Comment (pdunaj) [inline applications/nrf_desktop/src/hw_interface/Kconfig.buttons_sim:32]: `if DESKTOP_BUTTONS_SIM_ENABLE`

File: applications/nrf_desktop/src/hw_interface/Kconfig.buttons_sim
Comment (pdunaj) [inline applications/nrf_desktop/src/hw_interface/Kconfig.buttons_sim:11]: `depends on !CAF_BUTTONS`

File: applications/nrf_desktop/src/hw_interface/Kconfig.buttons_sim
Comment (pdunaj) [reply applications/nrf_desktop/src/hw_interface/Kconfig.buttons_sim:32]: you can put all configs from line 13 onwards under the same if

File: applications/nrf_desktop/src/hw_interface/Kconfig.buttons_sim
Comment (pdunaj) [inline applications/nrf_desktop/src/hw_interface/Kconfig.buttons_sim:2]: date

File: applications/nrf_desktop/src/hw_interface/Kconfig.click_detector
Comment (pdunaj) [inline applications/nrf_desktop/src/hw_interface/Kconfig.click_detector:11]: this should actually depends on CAF_BUTTON_EVENTS` as we don't care what is the source of the button events - only that these events are delivered

Comment (pdunaj) [reply applications/nrf_desktop/src/hw_interface/Kconfig.click_detector:11]: the same for motion and passkey (potentially other places?)

File: applications/nrf_desktop/src/hw_interface/Kconfig.selector
Comment (pdunaj) [inline applications/nrf_desktop/src/hw_interface/Kconfig.selector:10]: why?

File: applications/nrf_desktop/src/hw_interface/Kconfig.buttons_sim
Comment (pdunaj) [reply applications/nrf_desktop/src/hw_interface/Kconfig.buttons_sim:11]: also `depends on CAF_BUTTON_EVENTS`

File: applications/nrf_desktop/src/modules/Kconfig.ble
Comment (pdunaj) [inline applications/nrf_desktop/src/modules/Kconfig.ble:35]: `depends on CAF_BUTTON_EVENTS`

File: applications/nrf_desktop/src/modules/Kconfig.profiler_sync
Comment (pdunaj) [inline applications/nrf_desktop/src/modules/Kconfig.profiler_sync:12]: ?

File: applications/nrf_desktop/src/modules/hid_state.c
Comment (pdunaj) [inline applications/nrf_desktop/src/modules/hid_state.c:1426]: `CONFIG_CAF_BUTTON_EVENTS`

note that we don't care what is the source of the event - only that something delivers it

File: applications/nrf_desktop/src/hw_interface/Kconfig.buttons_sim
Comment (pdunaj) [reply applications/nrf_desktop/src/hw_interface/Kconfig.buttons_sim:11]: or `select CAF_BUTTON_EVENTS`

File: subsys/caf/modules/buttons.c
Comment (pdunaj) [inline subsys/caf/modules/buttons.c:406]: I have a better idea:
```
EVENT_SUBMIT(new_wake_up_event());
```

Comment (pdunaj) [reply subsys/caf/modules/buttons.c:406]: `;` will not be needed

**Review by pdunaj — CHANGES_REQUESTED** [review]

File: applications/nrf_desktop/src/hw_interface/Kconfig.selector
Comment (emob-nordic) [reply applications/nrf_desktop/src/hw_interface/Kconfig.selector:10]: Because `selector_hw.c` includes files that are part of CAF. I didn't want to split those for now as this is just a porting PR.

File: applications/nrf_desktop/src/modules/Kconfig.profiler_sync
Comment (emob-nordic) [reply applications/nrf_desktop/src/modules/Kconfig.profiler_sync:12]: Same as for `Kconfig.selector`

File: applications/nrf_desktop/src/hw_interface/Kconfig.buttons_sim
Comment (emob-nordic) [reply applications/nrf_desktop/src/hw_interface/Kconfig.buttons_sim:11]: I think `depends on CAF_BUTTON_EVENTS` is the most secure way of doing so. In theory nothing prevents from configuring buttons and simulated buttons at once so it might be either simulated or actual buttons.

Comment (emob-nordic) [reply applications/nrf_desktop/src/hw_interface/Kconfig.buttons_sim:11]: In fact this configuration (real + simulated buttons) is used: `configuration/nrf52840dk_nrf52840/app_ZDebug_keyboard.conf`. Buttons sim is `nrf_desktop` specific and actually depends on `CAF_BUTTONS`, button press is used to start/stop simulation. I don't think `depends on CAF_BUTTONS` should be removed from here.

File: applications/nrf_desktop/src/hw_interface/Kconfig.click_detector
Comment (emob-nordic) [reply applications/nrf_desktop/src/hw_interface/Kconfig.click_detector:11]: Click detector is `nrf_desktop` specific (at least for now) and actually `depends on CAF_BUTTONS`. Loosening this dependency may lead to configuration issues.

Comment (pdunaj) [reply applications/nrf_desktop/src/hw_interface/Kconfig.click_detector:11]: it depends on the events being sent not the buttons
note that if we add dependency on caf_buttons then user won't be able to use buttons implementation from the application

File: applications/nrf_desktop/src/modules/Kconfig.ble
Comment (emob-nordic) [reply applications/nrf_desktop/src/modules/Kconfig.ble:35]: Similar as for `Kconfig.click_detector`.

File: applications/nrf_desktop/src/hw_interface/Kconfig.click_detector
Comment (pdunaj) [reply applications/nrf_desktop/src/hw_interface/Kconfig.click_detector:11]: when everything was inside the nrf_desktop it was slightly different as nrf_desktop as a whole knew about all possible implementations

File: applications/nrf_desktop/src/modules/hid_state.c
Comment (emob-nordic) [reply applications/nrf_desktop/src/modules/hid_state.c:1426]: Similar as for `Kconfig.click_detector`.

File: applications/nrf_desktop/src/hw_interface/Kconfig.click_detector
Comment (emob-nordic) [reply applications/nrf_desktop/src/hw_interface/Kconfig.click_detector:11]: Agree. Will change it for all similar places.

**Review by pdunaj — APPROVED** [review]

File: subsys/caf/Kconfig
Comment (greg-fer) [inline subsys/caf/Kconfig:13]: `Enable Common Application Framework for use with the application.`

**Comment by carlescufi** [comment]
> I have no better idea than this in the PR to resolve that issue. @carlescufi any ideas?

Sorry, was late to this. I assume you fixed it @emob-nordic

**Comment by emob-nordic** [comment]
> > I have no better idea than this in the PR to resolve that issue. @carlescufi any ideas?
> 
> Sorry, was late to this. I assume you fixed it @emob-nordic

I left it as is (last commit). Adding braces tricks compliance check and hides the error.

**Comment by tejlmand** [comment]
> > Hi @carlescufi , is the location fine with you? It could be in applications/caf but subsystems also looks nice.
> 
> `subsys/caf` is a good location, thanks for checking @pdunaj.

Should we maybe spell out the folder name: `subsys/common_application_framework` so that it's clearer to the reader what the content in that folder is ?

**Comment by emob-nordic** [comment]
> Should we maybe spell out the folder name: `subsys/common_application_framework` so that it's clearer to the reader what the content in that folder is ?

Im not a fan of this as this name is way too long. Similar abbreviation is made for `esb` which is Nordics proprietary protocol.

**Comment by tejlmand** [comment]
> > Should we maybe spell out the folder name: `subsys/common_application_framework` so that it's clearer to the reader what the content in that folder is ?
> 
> Im not a fan of this as this name is way too long. Similar abbreviation is made for `esb` which is Nordics proprietary protocol.

I'm not talking about expanding everywhere, just expanding it for the folder name.
How often do you type a full folder name, compared to `co<TAB>` ?

With regards to ESB, then that's a proprietary protocol that has been there for more than a decade (~2009 if I recall correctly), and thus a known abbreviation regarding Nordic products.

caf is a new abbreviation, not known to anyone.

**Review by greg-fer — APPROVED** [review]

**Comment by carlescufi** [comment]
I think `subsys/common_application_framework` is overly long. I missed this change in my first pass at this. I'd rather we remain consistent with the `Kconfig` options and move this to `subsys/caf`.

**Comment by carlescufi** [comment]
> > > Hi @carlescufi , is the location fine with you? It could be in applications/caf but subsystems also looks nice.
> > 
> > 
> > `subsys/caf` is a good location, thanks for checking @pdunaj.
> 
> Should we maybe spell out the folder name: `subsys/common_application_framework` so that it's clearer to the reader what the content in that folder is ?

I disagree, see my last comment.

**Comment by pdunaj** [comment]
> > > > Hi @carlescufi , is the location fine with you? It could be in applications/caf but subsystems also looks nice.
> > > 
> > > 
> > > `subsys/caf` is a good location, thanks for checking @pdunaj.
> > 
> > 
> > Should we maybe spell out the folder name: `subsys/common_application_framework` so that it's clearer to the reader what the content in that folder is ?
> 
> I disagree, see my last comment.

I don't care; personally I preferred caf but I understand @tejlmand reasoning.
But please don't block this PR - we need it to continue work on EI. The name can be fixed in subsequent PR without harm to anybody.

**Comment by tejlmand** [comment]
> But please don't block this PR - 

I'm not blocking it. I approved it 16 days ago.
image

and it's still approved by me, from what I can tell:
image

Just wanted to raise the question to see if other agrees.

**Comment by carlescufi** [comment]
> I'm not blocking it. I approved it 16 days ago.

In that case please do switch to `subsys/caf` before we merge.

**Comment by pdunaj** [comment]
> > But please don't block this PR -
> 
> I'm not blocking it. I approved it 16 days ago.
> image
> 
> and it's still approved by me, from what I can tell:
> image
> 
> Just wanted to raise the question to see if other agrees.

@tejlmand  I am not talking about you but @carlescufi  who blocked merging. I don't care what the name is but it can be discussed later on. Other person the Emil is blocked on this PR and I don't see how individual preferences of somebody who came late should block.

**Comment by pdunaj** [comment]
> > I'm not blocking it. I approved it 16 days ago.
> 
> In that case please do switch to `subsys/caf` before we merge.

@carlescufi This is what I meant by blocking. You are doing it just now for no good reason.

**Comment by carlescufi** [comment]
OK, fine. I will merge this and then you submit another PR for renaming it.

**Comment by pdunaj** [comment]
> OK, fine. I will merge this and then you submit another PR for renaming it.

will be done now

**Comment by tejlmand** [comment]
> @tejlmand I am not talking about you but @carlescufi who blocked merging. 

Aah, my misunderstanding. I think in general we should be better see names from new users perspective, but it should not block ourselves from moving forward.

And often when it comes to abbreviations, we are very familiar with the context behind the abb. (and this includes myself), so we may not notice it until late in the process.

Things being noticed late in the process will also be commented late. But better mentioning it late, than not mentioning at all.
(or having to change even later in follow-up work)

**Pull request merged**
